src/modules/flow/netctl/netctl.c

```diff
@@ -0,0 +1,227 @@
+/*
+ * This file is part of the Soletta Project
+ *
+ * Copyright (C) 2015 Intel Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "sol-flow/netctl.h"
+#include "sol-flow-internal.h"
+
+#include <sol-netctl.h>
+#include <sol-str-table.h>
+#include <sol-util-internal.h>
+#include <errno.h>
+
+struct list_service_data {
+    struct sol_ptr_vector service_list;
+    struct sol_flow_node *node;
+};
+
+struct connection_service_data {
+    struct sol_flow_node *node;
+    char *ap_name;
+    bool connect;
+};
+
+void
+network_service_cb(void *data, const struct sol_netctl_service *service)
+{
+    struct list_service_data *mdata = data;
+    struct sol_flow_node *node = mdata->node;
+    int r;
+    const char *name;
+
+    if (sol_netctl_service_get_state(service) ==
+        SOL_NETCTL_SERVICE_STATE_REMOVE) {
+        sol_ptr_vector_del_element(&mdata->service_list, service);
+        return;
+    }
+
+    name = sol_netctl_service_get_name(service);
+    r = sol_flow_send_string_packet(node,
+        SOL_FLOW_NODE_TYPE_NETCTL_LIST_SERVICES__OUT__OUT,
+        name);
+    if (r < 0)
+        SOL_WRN("Failed to send network service name: %s", name);
+
+    sol_ptr_vector_append(&mdata->service_list, service);
+}
```

Comment:
I think we don't need to keep this mirror list `service_list`, since that's the same as using the `sol_netctl_get_services()`. However, that is not as dynamic as I'd expect.

What if we change `netclt/list-services` to be dynamic, on start it will list all existing services (`sol_netctl_get_services()`) and output then on `ADDED` port (instead of `OUT` port). Then when service is removed, you do emit its name in the `REMOVED` port, if a new service appears, then it would also go in the `ADDED` port.

Even in this case, there is no need to keep the mirror/shadow `mdata->services_list`.

What do you think? @bdilly @vcgomes @wzhen12